File: CompetitiveProgramming/homework/02/hw_1.py
# https://www.hackerrank.com/contests/usp2019-week-2/challenges
from sys import stdin
from typing import *

dist: Dict[Tuple[int, bool], int] = {}

def best_dist(arr, i, s, is_connected):
    if i == len(arr) - 2:
        return s + arr[i + 1] - arr[i]

    if (i, is_connected) in dist:
        return dist[(i, is_connected)]

    if is_connected:
        connected = best_dist(arr, i + 1, s + arr[i + 1] - arr[i], True)
        not_connected = best_dist(arr, i + 1, s, False)

        best = min(connected, not_connected)
    else:
        best = best_dist(arr, i + 1, s + arr[i + 1] - arr[i], True)

    dist[(i, is_connected)] = best
    return best

def main():
    num_of_nails = int(stdin.readline())
    nails = list(map(int, stdin.readline().split()))
    nails.sort()

    print(best_dist(nails, 0, 0, False))
    print(dist)


if __name__ == "__main__":
    main()

File: CompetitiveProgramming/homework/02/hw_1.cpp
// https://www.hackerrank.com/contests/usp2019-week-2/challenges
#include <bits/stdc++.h>
using namespace std;

int *positions;

int best_dist(int *arr, int i, int s, bool is_connected, int len) {
    if (i == len -2) {
        return s + arr[i + 1] - arr[i];
    }

    if (is_connected){
        return min(best_dist(arr, i + 1, s + arr[i + 1] - arr[i], true, len), best_dist(arr, i + 1, s, false, len));
    }
    else
    {
        return best_dist(arr, i + 1, s + arr[i + 1] - arr[i], true, len);
    }
}

int main()
{
    int num_of_nails;
    cin >> num_of_nails;

    int positions[num_of_nails];
    for (int i = 0; i < num_of_nails;++i) {
        cin >> positions[i];
    }

    sort(positions, positions+num_of_nails);

    int Solve(int idx){
        if(idx < 0) return 0;
        if(idx == 0) return 1E9;
        return min(Solve(idx - 2) + positions[idx] - positions[idx - 1], Solve(idx - 1) + positions[idx] - positions[idx - 1]);
    };

    cout << Solve(0) << endl;
    return 0;
}